### ISSUE TITLE
Making RPC -> ULT creation more robust

### ISSUE THREAD

**Issue by mdorier** [mochi-hpc/mochi-margo#186]
In [this macro](https://github.com/mochi-hpc/mochi-margo/blob/main/include/margo.h#L1529), we call `ABT_thread_create` to create the ULT that will run the body of an RPC. When scaling to tens of thousands of clients all hammering the same server, `ABT_thread_create` may fail with an error code of `ABT_ERR_MEM` (running out of memory), in particular because Margo sets the ULT stack size to 2MB.

We need to find a way to make this more robust, namely, the client should either receive an error back, or the creation of the ULT should be delayed until the ULT can actually be created.

Ideally, we would need both: throttling RPCs should not be done "just because `ABT_thread_create` failed", we should be able to have control over when and why it's done. We should allow the user to set a limit on the number of concurrent RPCs, and if `ABT_thread_create` fails we would return an error to the user, not throttle (in particular because we have no way of knowing exactly why `ABT_thread_create` failed. Maybe the server has a memory leak that filled up the memory? Maybe it holds large datasets that use a big portion of RAM? etc.)

Some thoughts on solutions:

I think we could implement a mechanism that adds an `int32_t` at the beginning of any response. This int would be used only so that margo can call `HG_Respond` without having any idea of the semantics of the actual response. I think I can do that the same way I implemented the injection of templated serialization code in thallium. This would solve the "sending an error back to the client" piece. We can then send errors like EAGAIN to the client when the number of RPCs exceeds a threshold.

**Comment by carns** [comment]
We've also intended for a while to have a custom stack allocator: https://github.com/mochi-hpc/mochi-margo/issues/166

I wonder if we did this if there is a way to have Argobots defer once some memory cap is reached (rather than allocating more stacks indefinitely).  It seems like it should be able to defer allocating a stack until memory is available.  We might also need to use the PRIO_WAIT pool to favor completing existing ULTs before starting execution of new ones.  I'm not sure how that fits together.  I'll look into the Argobots side of this.

We might need both eventually (ability to throttle at upper level because of some other resource, possibly unknown to Argobots, and ability to throttle thread creation in Argobots, because not all threads are known to Margo).
